File: android/modules/api/src/ti/modules/titanium/api/APIModule.java
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2009-2010 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */
package ti.modules.titanium.api;

import org.appcelerator.kroll.KrollModule;
import org.appcelerator.kroll.annotations.Kroll;
import org.appcelerator.titanium.TiContext;
import org.appcelerator.titanium.util.Log;
import org.appcelerator.titanium.util.TiLogger;

@Kroll.module
public class APIModule extends KrollModule
{
	private static final String LCAT = "TiAPI";

	@Kroll.constant public static final int TRACE = TiLogger.TRACE;
	@Kroll.constant public static final int DEBUG = TiLogger.DEBUG;
	@Kroll.constant public static final int INFO = TiLogger.INFO;
	@Kroll.constant public static final int NOTICE = TiLogger.NOTICE;
	@Kroll.constant public static final int WARN = TiLogger.WARN;
	@Kroll.constant public static final int ERROR = TiLogger.ERROR;
	@Kroll.constant public static final int CRITICAL = TiLogger.CRITICAL;
	@Kroll.constant public static final int FATAL = TiLogger.FATAL;

	public APIModule()
	{
		super();
	}

	public APIModule(TiContext tiContext)
	{
		this();
	}

	private TiLogger logger = new TiLogger()
	{
		public void log(int level, String message)
		{
			internalLog(level, message);
		}
	};

	public TiLogger getLogger()
	{
		return logger;
	}

	public void setLogger(TiLogger logger)
	{
		this.logger = logger;
	}

	private String toString(Object msg)
	{
		if (msg == null) {
			return "null";
		}
		return msg.toString();
	}
	
	@Kroll.method
	public void debug(Object msg)
	{
		logger.log(DEBUG, toString(msg));
	}

	@Kroll.method
	public void info(Object msg)
	{
		logger.log(INFO, toString(msg));
	}

	@Kroll.method
	public void warn(Object msg)
	{
		logger.log(WARN, toString(msg));
	}

	@Kroll.method
	public void error(Object msg)
	{
		logger.log(ERROR, toString(msg));
	}

	@Kroll.method
	public void trace(Object msg)
	{
		logger.log(TRACE, toString(msg));
	}

	@Kroll.method
	public void notice(Object msg)
	{
		logger.log(NOTICE, toString(msg));
	}

	@Kroll.method
	public void critical(Object msg)
	{
		logger.log(CRITICAL, toString(msg));
	}

	@Kroll.method
	public void fatal(Object msg)
	{
		logger.log(FATAL, toString(msg));
	}

	@Kroll.method
	public void log(String level, Object msg)
	{
		String ulevel = level.toUpperCase();
		int severity = INFO;

		if ("TRACE".equals(ulevel)) {
			severity = TRACE;
		} else if ("DEBUG".equals(ulevel)) {
			severity = DEBUG;
		} else if ("INFO".equals(ulevel)) {
			severity = INFO;
		} else if ("NOTICE".equals(ulevel)) {
			severity = NOTICE;
		} else if ("WARN".equals(ulevel)) {
			severity = WARN;
		} else if ("ERROR".equals(ulevel)) {
			severity = ERROR;
		} else if ("CRITICAL".equals(ulevel)) {
			severity = CRITICAL;
		} else if ("FATAL".equals(ulevel)) {
			severity = FATAL;
		} else {
			msg = "[" + level + "] " + msg;
		}

		logger.log(severity, toString(msg));
	}

	public void internalLog(int severity, String msg)
	{
		if (severity == TRACE)
		{
			Log.v(LCAT,msg);
		}
		else if (severity < INFO)
		{
			Log.d(LCAT,msg);
		}
		else if (severity < WARN)
		{
			Log.i(LCAT,msg);
		}
		else if (severity == WARN)
		{
			Log.w(LCAT,msg);
		}
		else
		{
			Log.e(LCAT,msg);
		}
	}
}

File: iphone/Classes/APIModule.m
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2009-2010 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */
#import "APIModule.h"
#import "TiApp.h"
#import "TiBase.h"
#import "TiExceptionHandler.h"
#import "TiUtils.h"

#ifndef USE_JSCORE_FRAMEWORK
#import "TiDebugger.h"
#endif

extern NSString *const TI_APPLICATION_DEPLOYTYPE;

@implementation APIModule

- (NSString *)apiName
{
  return @"Ti.API";
}

- (void)logMessage:(NSArray *)args severity:(NSString *)severity
{

  NSString *lcSeverity = [severity lowercaseString];
#ifndef USE_JSCORE_FRAMEWORK
  DebuggerLogLevel level = OUT;
  if ([lcSeverity isEqualToString:@"warn"]) {
    level = WARN;
  } else if ([lcSeverity isEqualToString:@"error"] ||
      [lcSeverity isEqualToString:@"critical"] ||
      [lcSeverity isEqualToString:@"fatal"]) {
    level = ERR;
  } else if ([lcSeverity isEqualToString:@"trace"]) {
    level = TRACE;
  } else if ([lcSeverity isEqualToString:@"debug"]) {
    level = LOG_DEBUG;
  }

  if ([[TiApp app] debugMode]) {
    NSMutableArray *messages = [NSMutableArray arrayWithArray:args];

    if (![lcSeverity isEqualToString:@"info"]) { // Custom severity, or just a badly-formed log; either way, debugger treats it as info
      [messages insertObject:[NSString stringWithFormat:@"[%@]", severity] atIndex:0];
    }

    TiDebuggerLogMessage(level, [messages componentsJoinedByString:@" "]);
  } else
#endif
  {
#ifndef USE_JSCORE_FRAMEWORK
    if ([TI_APPLICATION_DEPLOYTYPE isEqualToString:@"production"]) {
      if (level != ERR) {
        return;
      }
    }
#endif
    NSLog(@"[%@] %@", [severity uppercaseString], [args componentsJoinedByString:@" "]);
  }
}

- (id)transform:(id)arg
{
  if ([arg isKindOfClass:[NSDictionary class]]) {
    return [[[[TiScriptError alloc] initWithDictionary:arg] autorelease] description];
  }
  return arg;
}

- (void)debug:(NSArray *)args
{
  [self logMessage:args severity:@"debug"];
}

- (void)info:(NSArray *)args
{
  [self logMessage:args severity:@"info"];
}

- (void)warn:(NSArray *)args
{
  [self logMessage:args severity:@"warn"];
}

- (void)error:(NSArray *)args
{
  [self logMessage:args severity:@"error"];
}

- (void)trace:(NSArray *)args
{
  [self logMessage:args severity:@"trace"];
}

- (void)timestamp:(NSArray *)args
{
  NSLog(@"[TIMESTAMP] %f %@", [NSDate timeIntervalSinceReferenceDate], [self transform:[args objectAtIndex:0]]);
  fflush(stderr);
}

- (void)notice:(NSArray *)args
{
  [self logMessage:args severity:@"info"];
}

- (void)critical:(NSArray *)args
{
  [self logMessage:args severity:@"error"];
}

- (void)log:(NSArray *)args
{
  if ([args count] > 1) {
    [self logMessage:[args subarrayWithRange:NSMakeRange(1, [args count] - 1)] severity:[args objectAtIndex:0]];
  } else {
    [self logMessage:args severity:@"info"];
  }
}

- (void)reportUnhandledException:(NSArray *)args
{
  id lineNumber = [args objectAtIndex:0];
  id source = [args objectAtIndex:1];
  id message = [args objectAtIndex:2];

  NSLog(@"[ERROR] %@:%@ %@", source, lineNumber, message);
  fflush(stderr);
}

@end
